File: iOS/Session/Sources/Session/CurrentAccount/CurrentAccountSnapshot.swift
import Foundation

/// The persisted form of the signed-in account: the bundle plus the moment the server
/// last confirmed it. `validatedAt == nil` means the snapshot has never been validated
/// against the authenticated current-account endpoint (for example, it was migrated from
/// the legacy raw-bundle format) and is therefore stale.
public struct CurrentAccountSnapshot: Codable, Equatable, Sendable {
    public let account: AccountBundle
    public let validatedAt: Date?

    public init(account: AccountBundle, validatedAt: Date?) {
        self.account = account
        self.validatedAt = validatedAt
    }
}

extension CurrentAccountSnapshot {
    /// Decodes the persisted account, accepting both the envelope format and the legacy
    /// raw `AccountBundle` written by previous releases. A legacy value is wrapped as
    /// never-validated so the next eligible trigger reconciles it, without signing the
    /// gambler out.
    static func decode(from raw: Data) throws -> CurrentAccountSnapshot {
        let decoder = JSONDecoder()
        if let snapshot = try? decoder.decode(CurrentAccountSnapshot.self, from: raw) {
            return snapshot
        }
        let legacyBundle = try decoder.decode(AccountBundle.self, from: raw)
        return CurrentAccountSnapshot(account: legacyBundle, validatedAt: nil)
    }
}

File: Android/session/src/main/java/com/felipearpa/tyche/session/CurrentAccountSnapshot.kt
package com.felipearpa.tyche.session

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Persisted cache envelope for the signed-in account.
 *
 * [validatedAtEpochMillis] records the last successful server validation. `null` means the
 * snapshot was never validated — either it predates validation (a migrated legacy bundle) or
 * it has not been reconciled yet — and is therefore treated as stale.
 */
@Serializable
data class CurrentAccountSnapshot(
    val account: AccountBundle,
    val validatedAtEpochMillis: Long? = null,
)

private val tolerantJson = Json { ignoreUnknownKeys = true }

/**
 * Decodes a persisted current-account payload, accepting both the envelope format and the
 * legacy raw [AccountBundle] that released versions persisted. A legacy bundle migrates as a
 * never-validated (stale) snapshot so existing signed-in gamblers are not logged out. An
 * unrecognizable payload yields `null` instead of throwing.
 */
internal fun decodeCurrentAccountSnapshot(raw: String): CurrentAccountSnapshot? =
    runCatching { tolerantJson.decodeFromString<CurrentAccountSnapshot>(raw) }
        .recoverCatching {
            CurrentAccountSnapshot(
                account = tolerantJson.decodeFromString<AccountBundle>(raw),
                validatedAtEpochMillis = null,
            )
        }
        .getOrNull()
